### PR TITLE
Keep transform results in immutable caches in memory between builds

### LINKED ISSUE #15736: Re-enable cross-build transform cache
Re-introduces some of the changes of #15643.

We only should keep the artifact transform caches in-memory between builds for transforms with outputs within Gradle's immutable caches. Artifacts outside of the immutable caches are not guaranteed to be there on the next build, so we can't keep those in the cross-build cache.

---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by wolfs**
Fixes #15736.

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyManagementGradleUserHomeScopeServices.java [hunk 1]
```diff
@@ -101,7 +102,10 @@ ImmutableTransformationWorkspaceServices createTransformerWorkspaceServices(
                 .withDisplayName("Artifact transforms cache"),
             fileAccessTimeJournal,
             executionHistoryStore,
-            crossBuildInMemoryCacheFactory.newCacheRetainingDataFromPreviousBuild(Try::isSuccessful)
+            crossBuildInMemoryCacheFactory.newCacheRetainingDataFromPreviousBuild(result -> result
+                .map(transformedFiles -> transformedFiles.stream().allMatch(transformedFile -> globalCacheLocations.isInsideGlobalCache(transformedFile.getAbsolutePath())))
```
Comment (lptr) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyManagementGradleUserHomeScopeServices.java:106]: Nit: break `.allMatch()` into new line?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/android/AbstractRealLifeAndroidBuildPerformanceTest.groovy [hunk 2]
```diff
@@ -28,7 +28,7 @@ class AbstractRealLifeAndroidBuildPerformanceTest extends AbstractCrossVersionPe
 
     def setup() {
         runner.args = [AndroidGradlePluginVersions.OVERRIDE_VERSION_CHECK]
-        runner.targetVersions = ["6.8-20201116230039+0000"]
+        runner.targetVersions = ["6.8-20201223103229+0000"]
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/android/AbstractRealLifeAndroidBuildPerformanceTest.groovy:31]: Why change this?

**Review by lptr — APPROVED** [review]
LGTM.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/android/AbstractRealLifeAndroidBuildPerformanceTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/android/AbstractRealLifeAndroidBuildPerformanceTest.groovy:31]: @blindpirate [reverted my revert](https://github.com/gradle/gradle/pull/15738/commits/18f3930a270005177543e86e1e8f7c32edd352f7) from `release` on `master`, and those changes were part of the revert of the revert of the revert. I wouldn't touch it now and instead rebaseline `master` after these changes have been merged to lock in the performance improvements.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyManagementGradleUserHomeScopeServices.java (hunk 1 above)
Comment (wolfs) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyManagementGradleUserHomeScopeServices.java]: I added a line break.

**Pull request merged**